### ISSUE TITLE
Write an interactive tutorial

### ISSUE THREAD

**Issue by lamblin** [mila-udem/fuel#195]
- [x] Make an ipython notebook (maybe some slides as well) for the Deep Learning Summer School 2015 (https://github.com/mila-udem/summerschool2015). Done in https://github.com/mila-udem/summerschool2015/pull/2
- [x] Adapt it and integrate into Fuel's documentation

**Comment by bartvm** [comment]
On this topic, I dug up an old e-mail I wrote a long time ago about IPython notebooks for documentation:

> Right. There's 3 options for examples: Python files, IPython notebooks, or in the Sphinx-generated rST documentation. I think it's important to be able to mix code examples with explanations, links to documentation, etc. so I prefer the Sphinx/rSt option right now, because it automatically gets unit-tested. In Pylearn2 half the IPython notebooks are broken because they're hard to cover by unit tests. They are also annoying to use with Git, it's hard to link to other parts of the documentation from within them. Lastly, hosting them is more of a pain.

> However, I think interactivity is important too, and it's too bad that are only static... IPython does have a relatively unknown feature that we could advertise: Using "%doctest_mode" you can copy-paste code from the docs straight into IPython, and it will automatically filter out the unnecessary whitspace and >>> signs.

> Ideally (but this is quite long-term thinking), I think it would be amazing to have a Sphinx extension that automatically converts certain docs into IPython notebooks (see e.g. https://github.com/shashi/rst2ipynb) so that any tutorial would just have a link "Download as IPython Notebook" at the top... One can dream! But at least the possibility of doing this makes me feel better about sticking with rST for now.

**Comment by vdumoulin** [comment]
I wouldn't mind porting the notebook to Sphinx/rSt after the summer school.

As a presentation tool I prefer the notebook to static documentation, as it allows to do live coding, but I agree that Sphinx/rSt is more suited to the online docs.

**Comment by nouiz** [comment]
Interesting to know all this. Thanks for sharing.
Le 13 juil. 2015 17:00, "Bart van Merriënboer" <notifications@github.com> a
écrit :

**Comment by dwf** [comment]
Incidentally... https://pypi.python.org/pypi/pytest-ipynb

I'm hearing more and more good things about pytest.

**Comment by fvisin** [comment]
Copied and pasting for loops from rst files result in an error in iPython, e.g.:
```python
In [87]: %doctest_mode
Exception reporting mode: Plain
Doctest mode is: ON
>>> >>> for scheme in schemes:
...     ...     print(list(scheme.get_request_iterator()))
...     
  File "<ipython-input-88-b8e161d1b3a2>", line 2
    ...     print(list(scheme.get_request_iterator()))
    ^
SyntaxError: invalid syntax
```
```python
>>> >>> state = dataset.open()
>>> >>> scheme = ShuffledScheme(examples=dataset.num_examples, batch_size=4)
>>> >>> for request in scheme.get_request_iterator():
...     ...     data = dataset.get_data(state=state, request=request)
...     ...     print(data[0].shape, data[1].shape)
...     
  File "<ipython-input-100-f0f3028f2a0e>", line 2
    ...     data = dataset.get_data(state=state, request=request)
```
Apparently the `. . .` is not filtered out correctly in the case of `for` loops. I didn't investigate this very much though. Any comment on this?

**Comment by lamblin** [comment]
Does that work if you use the `%paste` magic instead of explicitly pasting?

**Comment by fvisin** [comment]
@lamblin: Good catch, works with `%paste`. 

It's a good workaround, but it is not as convenient as pasting with Ctrl-V while you follow the tutorial.

**Comment by lamblin** [comment]
Pasting code with ctrl-v does not work reliably with Python prompts anyway, because of the automatic indentation that the prompt tries to make for you. This is probably why `%paste` exists...

**Comment by vdumoulin** [comment]
Fixed by #225.
